### ISSUE TITLE
When I query clickhouse over ClickHouse-Native-JDBC,I found “Garbled characters” for Chinese characters

### ISSUE THREAD

**Issue by wallace-clickhouse** [housepower/ClickHouse-Native-JDBC#152]
### JDBC version in pom.xml
2.2-stable

### ClickHouse version
20.3.2.1

### Error  logs
“Garbled characters” for Chinese characters in query result

### Steps to reproduce
The default CharSet is POSIX on my linux server which clickhouse server runs on.
When I import datafile including Chinese characters which is encoded by UTF-8 with clickhouse-client，I found I got “Garbled characters” for Chinese characters if I query clickhouse over ClickHouse-Native-JDBC.So could you give me some advise to sovle this problem? I have the following initial ideas:
1.Modify the  default CharSet of my linux server from POSIX to en_US.UTF-8, and I will have a try;
2.Can ClickHouse-Native-JDBC provide a configurable way for user to specify the  CharSet  which is independent with the host platform?
Thank you very much!
 
### Other descriptions
None

**Comment by tauitdnmd** [comment]
Could u please provides your sample dataset and your code in order to reproduce the error?

**Comment by wallace-clickhouse** [comment]
OK.
Create table test (aa Int,bb String) ENGINE=MergeTree() PARTITION BY aa ORDER BY aa SETTINGS index_granularity=8192;
insert into table test(1,"中国");
insert into table test(2,"韩国");
insert into table test(3,"俄罗斯");
If the default CharSet of Linux is POSIX,you will get  “Garbled characters” when querying over  ClickHouse-Native-JDBC.

**Comment by sundy-li** [comment]
Are you using `insert into table test(1,"中国");` ?

why not `insert into table test(1, '中国' );`

**Comment by wallace-clickhouse** [comment]
Yeah,you're right,the SQL statement should be 
insert into table test(1, '中国' );
the same problem occured.

**Comment by sundy-li** [comment]
@wallace-clickhouse  

Seems`integration-tests ` in #153  is ok, I can't reproduce it.

**Comment by wallace-clickhouse** [comment]
Firstly,you can set the ChrarSet of your linux server which clickhouse runs on to 'POSIX'.
Secondly,you can import the datafile formatted by csv into clickhouse with clickhouse-client command,the csv datafile's CharSet is UTF-8.
Then if you query the data importted with ClickHouse-Native-JDBC,you will get “Garbled characters” for Chinese characters.
please refer to the issuse #132  "Fix problems which makes the JDBC library not platform independent #132"
thank you

**Comment by wallace-clickhouse** [comment]
Hi,friends
I just want a JDBC charset configuration which is independent with the charset of host platform.
Have I explained my question clearly?

**Comment by sundy-li** [comment]
Yes, I see your points. Is there any error logs or strace logs? I still can't reproduce it.

**Comment by pan3793** [comment]
@wallace-clickhouse Would you please provided the OS version and glibc version? It may be related to glibc.

**Comment by uschindler** [comment]
> I just want a JDBC charset configuration which is independent with the charset of host platform.

That's already the case since #132. Since that commit, the charset used to en/decode data from/to Clickhouse is always UTF-8. Of course if you have an old database or your pushed the data with the native client into the database, it my not be in UTF-8 and so the characters get garbled.

**Comment by wallace-clickhouse** [comment]
@uschindler 
the native client,you mean clickhouse-client？

**Comment by uschindler** [comment]
I'd recommend to the reporter, to make sure that the following works. First some background:
- Clickhouse does not have any encoding in "String" datatype. For clickhouse it is just "bytes".
- The JDBC driver (currently) always assumes UTF-8 and we have integration tests and due to forbiddenapis this is ensured that we do not fallback to default charset when converting the underlying `byte[]` to a Java String. This was fixed in 2.2 already, PR #132 
- JDBC also allows to get the native bytes, use `ResultSet#getBytes(column)`, but unfortunately the JDBC driver does not implement it (let's open a PR)

As Clickhouse does not implement `ResultSet#getBytes(column)`, we cannot check here. But can you make sure that the bytes in database are really valid UTF-8?

IMHO, we should implement the missing ResultSet function, as this would allow to get the raw bytes, too. This also makes sense, because you can use "string" fields in clickhouse to store arbitrary BLOBs.

> the native client,you mean clickhouse-client？

Yes. If the locale is set to POSIX, without UTF-8, this one may wrongly parse the CSV file. So the data in clickhouse is already notUTF-8 encoded.

In short: Please import your data with UTF-8 CSV file using a native clickhouse-client running on a correct default locale. In the JDBC client you then should get the data nongarbled. The system locale has no effect on JDBC driver, it always uses UTF8. If this fails, the data in Clickouse is a different encoding.

**Comment by uschindler** [comment]
Nevertheless, I think it would be a good thing to allow to give a charset for the clickhouse encoding in the JDBC options. See #107. So people with incorrect data inside clickhouse can override the decoding of bytes in the JDBC client.

**Comment by wallace-clickhouse** [comment]
Yeah,I got the same conclusion with you.
Now when I import data into clickhouse with clickhouse-client,I must check the system locale first and make sure it's set as UTF8.
It works fine！

**Comment by uschindler** [comment]
I have the feeling that with CSV files, the clickhouse client uses some third party CSV parser which uses it's own idea of charset. I'd open an issue on their side.

We import our data using Google Protobuf with the clickhouse client and there everything works fine.

**Comment by uschindler** [comment]
I also opened #169 to implement get/setBytes() in ResultSet and PreparedStatement.

**Comment by pan3793** [comment]
@uschindler Cool! Thanks for your explanation.
